snow/engine/common/engine.go

```diff
@@ -502,3 +509,24 @@ type InternalHandler interface {
 	// Notify this engine of a message from the virtual machine.
 	Notify(context.Context, Message) error
 }
+
+// AppError is an application-defined error
+type AppError struct {
+	// Code is application-defined and should be used for error matching
+	Code int32
+	// Message is a human-readable error message
+	Message string
+}
+
+func (a *AppError) Error() string {
+	return fmt.Sprintf("%d: %s", a.Code, a.Message)
+}
+
+func (a *AppError) Is(target error) bool {
+	appErr, ok := target.(*AppError)
+	if !ok {
+		return false
+	}
+
+	return a.Code == appErr.Code
+}
```

Comment:
maybe it's over engineering, but I'd like this to be in its own errors.go file where we possibly list an error code enum if needed.